Add Internal Editorial for COCI Sjeckanje (#3009)
cpinitiative/usaco-guide: solutions/coci-21-sjeckanje.mdx

---
id: coci-21-sjeckanje
source: COCI 2021
title: Sjeckanje
author: Timothy Gao
---

> For any segment $[l, r]$ with more than three elements, if there exists a tuple such that $a_i < a_{i+2} < a_{i+1}$ or $a_i > a_{i+2} > a_{i+1}$ where $i...i+2$ are contained in the segment, then the segment can be split into two segments $l, l+1... i+1$ and $i+2, i+3, ... r$. This is provably at least as good as the score ($\text {max} - \text {min}$) of the initial segment. The tuples $a_i < a_{i+2} < a_{i+1}$ and $a_i > a_{i+2} > a_{i+1}$ must exist in any segment that is **not** non-increasing or non-decreasing, and the first two elements of the segment determines whether the rest of the segment will be non-increasing or non-decreasing. Therefore, an optimal solution can be constructed with only non-increasing or non-decreasing segments. This paves the way for several key properties.

Reviewer: > The tuples $a_i < a_{i+2} < a_{i+1}$ and $a_i > a_{i+2} > a_{i+1}$ must exist in any segment that is **not** non-increasing or non-decreasing,

this is not actually true. I think you should just say that the three elements aren't in strictly increasing or decreasing order.

> For any segment $[l, r]$ with more than three elements, if there exists a tuple such that $a_i < a_{i+2} < a_{i+1}$ or $a_i > a_{i+2} > a_{i+1}$ where $i...i+2$ are contained in the segment, then the segment can be split into two segments $l, l+1... i+1$ and $i+2, i+3, ... r$. This is provably at least as good as the score ($\text {max} - \text {min}$) of the initial segment. The tuples $a_i < a_{i+2} < a_{i+1}$ and $a_i > a_{i+2} > a_{i+1}$ must exist in any segment that is **not** non-increasing or non-decreasing, and the first two elements of the segment determines whether the rest of the segment will be non-increasing or non-decreasing. Therefore, an optimal solution can be constructed with only non-increasing or non-decreasing segments. This paves the way for several key properties.

Author: Ty for the suggestions.

Define array $D$ of length $N-1$ such that $D_i = A_{i+1} - A_i$. If the current segment $s = [l, r]$ is non-decreasing, then $\sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$ and $D_i \geq 0$ for $i \in [l, r]$. If the current segment $s = [l, r]$ is non-decreasing, then $-1 \cdot \sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$ and $D_i \leq 0$ for $i \in [l, r]$. Between any two adjacent selected segments $s = [a, b]$ and $w = [b+1, c]$, $D_b$ must not included in our answer. With this formulation, adding integer $x$ to $l, l+1, ... r$ is equivalent to performing the point updates $D_{l-1}$ += $x$ and $D_{r}$ -= $x$.

> Now, the problem becomes selecting segments on $D$ rather than the original $A$, where each segment contributes the absolute value of the sum of its elements to the final answer. Every two neighboring segments must be separated by at least one element (see $D_b$ above), and all elements in each segment must be either $\leq 0$ or $\geq 0$. One way to go from here is with a modified segment tree. Each node stores the maximum value of its segment for each of the four cases of taking or not taking the first or last element, as well as the border elements of its segment.  When merging, we must pay attention to only combine two segments together if the signs of their border elements to be combined are the same.

Reviewer: ```suggestion
Now, the problem becomes selecting segments on $D$ rather than the original $A$, where each segment contributes the absolute value of the sum of its elements to the final answer. Every two neighboring segments must be separated by at least one element (see $D_b$ above), and all elements in each segment must be either $\leq 0$ or $\geq 0$. One way to go from here is with a segment tree. Each node stores the maximum value of its segment for each of the four cases of taking or not taking the first or last element, as well as the border elements of its segment.  When merging, we must pay attention to only combine two segments together if the signs of their border elements to be combined are the same.
```

**Time Complexity**: $\mathcal O(N + Q\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

> struct node{

Reviewer: there seems to be some inconsistency with the spacing here (e.g., before curly braces)

Now, the problem becomes selecting segments on $D$ rather than the original $A$, where each segment contributes the absolute value of the sum of its elements to the final answer. Every two neighboring segments must be separated by at least one element (see $D_b$ above), and all elements in each segment must be either $< 0$ or $> 0$. One way to go from here is with a segment tree. Each node stores the maximum value of its segment for each of the four cases of taking or not taking the first or last element, as well as the border elements of its segment.  When merging, we must pay attention to only combine two segments together if the signs of their border elements to be combined are the same to ensure that the combined segment continues to be either strictly increasing or decreasing.

**Time Complexity**: $\mathcal O(N + Q\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

> struct node

Reviewer 2: ```suggestion

struct node
```
also rename to `Node`

{
	ll borders[2] = {}; // border values of segment

> 	ll dp[2][2] = {};   // take/not-take left and right border values

Reviewer 2: ```suggestion
	ll borders[2] = {};  // border values of segment
	ll dp[2][2] = {};  // take/not-take left and right border values

	node() {}
	node(ll v)
	{
		borders[0] = borders[1] = v;
		dp[0][1] = dp[1][0] = dp[0][0] = 0;
		dp[1][1] = abs(v);
	}

> 	// assume current is left

Reviewer 2: space out your functions

Author: Change made

	node comb(node &other)
	{
		node ret;
		ret.borders[0] = borders[0];
		ret.borders[1] = other.borders[1];
		// [	 ][	 ]
		// l	 mo	 r
		// m and o same -> taken segments are combined
		for (int l = 0; l < 2; l++)
		{
			for (int m = 0; m < 2; m++)
			{
				for (int o = 0; o < 2; o++)
				{
					for (int r = 0; r < 2; r++)

> 					{

Reviewer 3: the nested loops are insane!

Author: Yea... I mean it's [pretty common](https://usaco.guide/problems/poi-2014cards/solution) for merging segment tree nodes. Is there anything I should fix?

						if (m && o)
						{
							// it's never optimal to take two opposite sign values
							if ((borders[1] < 0) == (other.borders[0] < 0))
							{
								ret.dp[l][r] = max(ret.dp[l][r], dp[l][m] + other.dp[o][r]);
							}
						}
						else
						{
							ret.dp[l][r] = max(ret.dp[l][r], dp[l][m] + other.dp[o][r]);
						}
					}
				}
			}
		}
		return ret;
	}
	// modify a single-element node segment by +v
	void upd(ll v)
	{
		borders[0] += v;
		borders[1] += v;
		dp[1][1] = abs(borders[0]);
	}
};

> struct segtree

Reviewer 2: ```suggestion

struct segtree
```
rename to `SegTree`

{
	node val;
	int gL, gR, mid;
	segtree *left, *right;
	segtree(int l, int r, vector<node> &nums)
	{
		gL = l;
		gR = r;
		mid = (gL + gR) / 2;
		if (l == r)
		{
			val = nums[l];
		}
		else
		{
			left = new segtree(l, mid, nums), right = new segtree(mid + 1, r, nums);
			val = left->val.comb(right->val);
		}
	}
	void update(int idx, ll updval)
	{
		if (gL == gR)
		{
			val.upd(updval);
		}
		else
		{
			if (idx <= (gL + gR) / 2)
			{
				left->update(idx, updval);
			}
			else
			{
				right->update(idx, updval);
			}
			val = left->val.comb(right->val);
		}
	}
};
int main()
{
	ios_base::sync_with_stdio(false);

> 	cin.tie(0);

Reviewer 2: remove if runs w/o

Author: Needed for TL

	int N, Q;
	cin >> N >> Q;
	vector<node> D(N - 1);
	int a;
	cin >> a;
	// compute difference array
	for (int i = 0; i < N - 1; i++)
	{
		int b;
		cin >> b;
		D[i] = node(b - a);
		swap(a, b);
	}
	// a_(i + 1) - a(i)
	segtree sgt(0, N - 2, D);
	for (int q = 0; q < Q; q++)
	{
		int l, r;
		ll x;
		cin >> l >> r >> x;

> 		l--, r--;

Reviewer 2: separate lines

Author: Change made

**Time Complexity**: $\mathcal O(N + Q\log N)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

struct Node
{
	ll borders[2] = {};  // border values of segment
	ll dp[2][2] = {};  // take/not-take left and right border values
	Node() {}
	Node(ll v)
	{
		borders[0] = borders[1] = v;
		dp[0][1] = 0;
		dp[1][0] = 0;
		dp[0][0] = 0;
		dp[1][1] = abs(v);
	}
	// assume current is left
	Node comb(Node &other)
	{
		Node ret;
		ret.borders[0] = borders[0];
		ret.borders[1] = other.borders[1];
		// [	 ][	 ]
		// l	 mo	 r
		// m and o same -> taken segments are combined
		for (int l = 0; l < 2; l++)
		{
			for (int m = 0; m < 2; m++)
			{
				for (int o = 0; o < 2; o++)
				{
					for (int r = 0; r < 2; r++)
					{
						if (m && o)
						{
							// it's never optimal to take two opposite sign values
							if ((borders[1] < 0) == (other.borders[0] < 0))
							{
								ret.dp[l][r] = max(ret.dp[l][r], dp[l][m] + other.dp[o][r]);
							}
						}
						else
						{
							ret.dp[l][r] = max(ret.dp[l][r], dp[l][m] + other.dp[o][r]);
						}
					}
				}
			}
		}
		return ret;

> 	}
> 	// modify a single-element Node segment by +v

Reviewer 2: ```suggestion
	}

	// modify a single-element Node segment by +v
```

	void upd(ll v)
	{
		borders[0] += v;
		borders[1] += v;
		dp[1][1] = abs(borders[0]);
	}
};

struct SegTree
{
	Node val;
	int gL, gR, mid;
	SegTree *left, *right;
	SegTree(int l, int r, vector<Node> &nums)
	{
		gL = l;
		gR = r;
		mid = (gL + gR) / 2;
		if (l == r)
		{
			val = nums[l];
		}
		else
		{
			left = new SegTree(l, mid, nums), right = new SegTree(mid + 1, r, nums);
			val = left->val.comb(right->val);
		}

> 	}
> 	void update(int idx, ll updval)

Reviewer 2: ```suggestion
	}

	void update(int idx, ll updval)
```

	{
		if (gL == gR)
		{
			val.upd(updval);
		}
		else
		{
			if (idx <= (gL + gR) / 2)
			{
				left->update(idx, updval);
			}
			else
			{
				right->update(idx, updval);
			}
			val = left->val.comb(right->val);
		}
	}

> };
> int main()

Reviewer 2: ```suggestion
};

int main()
```

---
id: coci-21-sjeckanje
source: COCI 2021
title: Sjeckanje
author: Timothy Gao
---

> For any segment $[l, r]$ with more than three elements, if there exists a tuple such that $a_i \leq a_{i+2} \leq a_{i+1}$ or $a_i \geq a_{i+2} \geq a_{i+1}$ where $i$ to $i+2$ are contained in the segment, then the segment can be split into two segments $l, l+1, \dots, i+1$ and $i+2, i+3, \dots, r$. This is provably at least as good as the score ($\text {max} - \text {min}$) of the initial segment. If the segment isn't strictly increasing or strictly decreasing, there must be three consecutive elements within the segment that aren't in strictly increasing or decreasing order, and the first two elements of the segment determine whether the rest of the segment will be strictly increasing or decreasing. Therefore, an optimal solution can be constructed with only monotonically increasing or decreasing segments. This paves the way for several key properties.

Reviewer: ```suggestion
For any segment $[l, r]$ with at least three elements, if there exists a tuple such that $a_i \leq a_{i+2} \leq a_{i+1}$ or $a_i \geq a_{i+2} \geq a_{i+1}$ where $i$ to $i+2$ are contained in the segment, then the segment can be split into two segments $l, l+1, \dots, i+1$ and $i+2, i+3, \dots, r$. This is provably at least as good as the value of the initial segment ($\text {max} - \text {min}$). More generally, if a segment isn't strictly increasing or strictly decreasing, we can split it into two segments without lowering the total value. Therefore, an optimal solution can be constructed with only strictly increasing or decreasing segments. This paves the way for several key properties.
```

Reviewer: monotonic != strict btw https://en.wikipedia.org/wiki/Monotonic_function#In_calculus_and_analysis

Author: My argument initially was that the solution could be constructed with monotonously non-decreasing or non-increasing segments, but just having it be strictly increasing or decreasing makes the explanation more clear. Thanks for the suggestions.

> Define array $D$ of length $N-1$ such that $D_i = A_{i+1} - A_i$. If the current segment $s = [l, r]$ is non-decreasing, then $\sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$ and $D_i \geq 0$ for $i \in [l, r]$. If the current segment $s = [l, r]$ is non-decreasing, then $-1 \cdot \sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$ and $D_i \leq 0$ for $i \in [l, r]$. Between any two adjacent selected segments $s = [a, b]$ and $w = [b+1, c]$, $D_b$ must not included in our answer. With this formulation, adding integer $x$ to $l, l+1, ... r$ is equivalent to performing the point updates $D_{l-1}$ += $x$ and $D_{r}$ -= $x$.

Reviewer: ```suggestion
Define array $D$ of length $N-1$ such that $D_i = A_{i+1} - A_i$. If the current segment $s = [l, r]$ is non-decreasing, then $D_i \geq 0$ for $i \in [l, r)$ and $\sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$. If the current segment $s = [l, r]$ is non-increasing, then $D_i \leq 0$ for $i \in [l, r)$ and $-1 \cdot \sum_{i=l}^{r-1} D_i = \text {max}(s) - \text {min}(s)$. Between any two adjacent selected segments $s = [a, b]$ and $w = [b+1, c]$, $D_b$ must not included in our answer. With this formulation, adding integer $x$ to $l, l+1, \dots, r$ is equivalent to performing the point updates $D_{l-1}\mathrel{+}=x$ and $D_{r}\mathrel{-}=x$.
```
